src/solvers/smt2/smt2_conv.cpp

```diff
@@ -4390,6 +4403,27 @@ void smt2_convt::find_symbols(const exprt &expr)
       defined_expressions[expr]=id;
     }
   }
+  else if(expr.id() == ID_array_comprehension)
+  {
+    if(defined_expressions.find(expr) == defined_expressions.end())
+    {
+      const auto &array_comprehension = to_array_comprehension_expr(expr);
+
+      const irep_idt id =
+        "array_comprehension." + std::to_string(defined_expressions.size());
+      out << "(define-fun " << id << " () ";
```

Comment:
Why are you creating a function with no arguments that returns a lambda?

Reply:
Unless I have got the wrong end of the stick completely, I think you can implement this without the lambda.

Reply:
Arrays are infinite maps in SMT and are represented just like key -> value functions. The function with no arguments returns a lambda, because we want to return an array.

This has been tested with both Z3 and CVC4 -- you can run the attached regression test with `cbmc --smt2`.

I am not super sure about the official status of `lambda` within SMTLIB2 vs SMTLIB3, but both Z3 and CVC4 support it for a while now.
AFAIK, the only other way to implement array comprehension would be to use quantifiers, but that did not work even for the simpler case of `array_of` where the lambda expression was a constant (independent of the index).

Reply:
I will make this translation solver-specific then -- only emit SMT2 output if solver is either Z3 or CVC4.